File: java/src/main/java/workshop/Starter.java
package workshop;

import io.temporal.api.common.v1.WorkflowExecution;
import io.temporal.client.WorkflowClient;
import io.temporal.client.WorkflowOptions;
import io.temporal.serviceclient.WorkflowServiceStubs;
import java.time.Duration;

public class Starter {

    public static void main(String[] args) {
        String type = args.length > 0 ? args[0] : "hello";

        WorkflowServiceStubs service = WorkflowServiceStubs.newLocalServiceStubs();
        WorkflowClient client = WorkflowClient.newInstance(service);

        WorkflowExecution execution;
        String id;

        if (type.equals("multiactivity")) {
            int a = args.length > 1 ? Integer.parseInt(args[1]) : 3;
            int b = args.length > 2 ? Integer.parseInt(args[2]) : 4;
            id = args.length > 3 ? args[3] : "demo-wf";

            WorkflowOptions options =
                WorkflowOptions.newBuilder()
                    .setWorkflowId(id)
                    .setTaskQueue(WorkerApp.TASK_QUEUE)
                    .setWorkflowTaskTimeout(Duration.ofMinutes(15))
                    .build();

            PipelineWorkflow stub = client.newWorkflowStub(PipelineWorkflow.class, options);
            execution = WorkflowClient.start(stub::run, new MathInput(a, b));
        } else {
            String name = args.length > 1 ? args[1] : "Temporal";
            id = args.length > 2 ? args[2] : "demo-wf";

            WorkflowOptions options =
                WorkflowOptions.newBuilder()
                    .setWorkflowId(id)
                    .setTaskQueue(WorkerApp.TASK_QUEUE)
                    // 15-min workflow-task timeout gives you headroom to sit on a
                    // breakpoint in WORKFLOW code. Requires the server-side cap
                    // raised to 15m (see workshop/README.md).
                    .setWorkflowTaskTimeout(Duration.ofMinutes(15))
                    .build();

            switch (type) {
                case "signal": {
                    ApprovalWorkflow stub =
                        client.newWorkflowStub(ApprovalWorkflow.class, options);
                    execution = WorkflowClient.start(stub::run, new GreetingInput(name));
                    break;
                }
                case "nondet": {
                    NonDeterminismWorkflow stub =
                        client.newWorkflowStub(NonDeterminismWorkflow.class, options);
                    execution = WorkflowClient.start(stub::run, new GreetingInput(name));
                    break;
                }
                case "versioned": {
                    VersionedWorkflow stub =
                        client.newWorkflowStub(VersionedWorkflow.class, options);
                    execution = WorkflowClient.start(stub::run, new GreetingInput(name));
                    break;
                }
                case "hello":
                default: {
                    GreetingWorkflow stub =
                        client.newWorkflowStub(GreetingWorkflow.class, options);
                    execution = WorkflowClient.start(stub::greet, new GreetingInput(name));
                    break;
                }
            }
        }

        System.out.println(
            "Started " + type + " WorkflowID=" + id + " RunID=" + execution.getRunId());
        System.exit(0);
    }
}

File: python/starter.py
import argparse
import asyncio
from datetime import timedelta

from temporalio.client import Client

from workflows import (
    GreetingWorkflow,
    PipelineWorkflow,
    ApprovalWorkflow,
    NonDeterminismWorkflow,
    VersionedWorkflow,
)
from worker import TASK_QUEUE
from shared import GreetingInput, MathInput

# Map a workflow key to the class run method to start.
WORKFLOWS = {
    "hello": GreetingWorkflow.run,
    "multiactivity": PipelineWorkflow.run,
    "signal": ApprovalWorkflow.run,
    "nondet": NonDeterminismWorkflow.run,
    "versioned": VersionedWorkflow.run,
}


async def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--workflow", default="hello", choices=list(WORKFLOWS))
    parser.add_argument("--id", default="demo-wf")
    parser.add_argument("--name", default="Temporal")
    parser.add_argument("--a", type=int, default=3)
    parser.add_argument("--b", type=int, default=4)
    parser.add_argument("--wait", action="store_true")
    args = parser.parse_args()

    run = WORKFLOWS[args.workflow]

    client = await Client.connect("localhost:7233", namespace="default")

    arg = MathInput(a=args.a, b=args.b) if args.workflow == "multiactivity" else GreetingInput(name=args.name)

    handle = await client.start_workflow(
        run,
        arg,
        id=args.id,
        task_queue=TASK_QUEUE,
        # 15-min workflow-task timeout gives you headroom to sit on a breakpoint
        # in WORKFLOW code. Requires the server-side cap raised to 15m.
        task_timeout=timedelta(minutes=15),
    )

    run_id = handle.result_run_id or handle.first_execution_run_id
    print(f"Started {args.workflow} WorkflowID={args.id} RunID={run_id}")

    if args.wait:
        print("Result:", await handle.result())
    else:
        print("Signal-based demos need a signal to finish — see scripts/signal.sh")


if __name__ == "__main__":
    asyncio.run(main())
